debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java

```diff
@@ -315,12 +325,10 @@ public int getOperation() throws SQLException {
         }
 
         public Object[] getData() throws SQLException {
-            final int dataColumnCount = resultSet.getMetaData().getColumnCount() - (COL_DATA - 1);
-            final Object[] data = new Object[dataColumnCount];
-            for (int i = 0; i < dataColumnCount; i++) {
-                data[i] = resultSet.getObject(COL_DATA + i);
+            if (rowMapper == null) {
```

Comment:
Ok, that makes sense. In that light I think it'd be better then to maintain the row mapper at the level of `ChangeTable` itself. After all, `ChangeTable` represents a given capture instance with a given schema. This would avoid the lazy logic and make things a bit easier to grasp overall.

Reply:
Lazy logic is still needed. `ChangeTable`s are created even earlier than `ChangeTablePointer`.

Reply:
But ``ChangeTable``s have one fixed schema, don't they? I.e. you should be able to retrieve the equivalent row mapper right when instantiating a `ChangeTable`. Or am I missing something here?

Reply:
As far as I understand `ChangeTable` has only a reference `TableId sourceTableId`. Table schema is kept in `SqlServerDatabaseSchema` and can be retrieved using `TableId`.
A new schema is registered when a switch from the old capture instance to the new one occurs. The new schema overwrites the existing schema in `SqlServerDatabaseSchema`.

Reply:
Yes, so on that one it's still a bit unclear to my why the schema is historized here to begin with. I've filed [DBZ-1096](https://issues.jboss.org/browse/DBZ-1096) to get that clarified (CC @jpechane). That said, take a look at `migrateTable()`: it obtains the `Table` (=structural description) for a `ChangeTable` via `connection#getTableSchemaFromTable()`. It should be possible to do that when creating the `ChangeTable` itself and then bake the `Table` or just row mapper right into it.

I suppose there might be the case where a change table actually "changes" by dropping and recreating it using the same name with a diffent definition, but that seems like a rather contrived corner case which I think we can ignore.

Reply:
I've change the PR a bit. Table is set in ChangeTable just after its creation.

In the end I haven't moved ResultSetMapper to ChangeTable. It looks odd when part of ResultSet is interpreted in ChangeTablePointer ("system columns") and a part in ChangeTable (data columns).

Reply:
Eh, I overlooked one failing test.
The current code (creating resultSetMapper when ChangeTablePointer is created) does not work in one scenario - renaming column. This one is specific - there are no two capture instances at the same time. Renaming column required to disable cdc, which results in dropping capture instance as well, and then enabling cdc again...

Reply:
Could you add a new JIRA issue for that one? Thanks!

Reply:
I believe it (special handling of column rename) is not a bug, it's a feature. I need to change the code to handle it somehow.

Reply:
@gunnarmorling I'm sorry it takes so long. I was trying to avoid lazy initialization of the mapper, but I failed. To be honest I'm still reluctant to make bigger changes in the code since I do not know it very well.

The best I can provide now is to call `setSourceTable()` in appropriate places (`io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource#getCdcTablesToQuery` and `io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource#migrateTable`) and create resultSetMapper lazily :(